### ISSUE TITLE
Format decimal values in an activeForm field

### ISSUE THREAD

**Issue by gpoehl** [yiisoft/yii2#8406]
As far as I can see there are basicly two ways to format a decimal value for activeForm fields.

First one is to create getter und setter functions to prepare and return a formatted value. Here you can use the php NumberFormatter class which will use the language depended decimal sign and group separator. This way you need to write these functions for each decimal field (or one magic method) . 

The second way is using the maskedInput widget. Here you need to define the groupSeparator and radixPoint manually. I propose to set the default values for those fields within the widget based on the language.

A third option could be implemented similar to the widget function of the activeField class. You might allow using the formatter class or any other function which formats the output string.

**Comment by lynicidn** [comment]
`Yii::$app->formatter->asDecimal`

**Comment by twinni** [comment]
@lynicidn, that's the way to format decimals in normal views. But I don't see how to use this in combination with **activeForm fields**? 
```php
 <?= $form->field($model, 'ablesewert') ?>
```
Did I missed something?

**Comment by lynicidn** [comment]
Field should set value w/o prepare

**Comment by lynicidn** [comment]
or u can set value in manual mode

**Comment by twinni** [comment]
@lynicidn, the field need to be prepared. Decimal comma is an absolute requirement (in some countries). Beside that I don't know how to set value in manual mode. Or do you mean without active field widget? Can you pls. give an example?

In general I would expect some assistance from a framework or at least a clear documentation how to handle decimal values.

I already raised this question in the forum but got no response. All I found so far are some articles based on yii version 1. But nowhere an descripiton how to handle the whole process from reading a decimal value, render it (localized by country), check the rules and save the new value.

**Comment by lynicidn** [comment]
```php
$form->field($model, 'attr', ['inputOptions' => ['value' => Yii::$app->formatter->asDecimal($model->attr)]])
```
п.с. ну и извращенец вы батенька

**Comment by groehl** [comment]
Hi, I'm new to Yii2, have the same Problem and did it this way:
In model, i change the format bevor use:
```
    $this->Preis = yii::$app->formatter->asDecimal($this->Preis,2);
```
To write back, the input-value will be transformed by filter rule:
```
    ['Preis', 'filter', 'filter' => function ($value) {$value = str_replace(',', '.', $value); return $value; }],
```
It's not elegant and you have to do it in each model for every decimal, but it works.
Comments?

**Comment by cebe** [comment]
we could add a `format` property to ActiveField like we have it in the gridview columns.

**Comment by groehl** [comment]
... that would be great! It seems to be more like "dataFormat" or similar, because it should be adjustable for both directions.

**Comment by freezy-sk** [comment]
@cebe I could try to make a PR

**Comment by samdark** [comment]
@freezy-sk you're welcome.

**Comment by lynicidn** [comment]
@cebe better add setter and getter prioritet upper that getAttribute and setAttribute =)

**Comment by cebe** [comment]
@lynicidn AR should not be responsible for the presentation layer, you may want to get human readable format but also machine readable format for one attribute dependend on the context. thats why getters and setter are not good in this case.

**Comment by lynicidn** [comment]
@cebe i understand, in core need replace all $this->$name to $this->getAttribute($name) :(

**Comment by cebe** [comment]
what? why?

**Comment by lynicidn** [comment]
@cebe 
in rest model i can prepare value - `fields`
in base model too - getter and setter and remove public $attribute property from model
in ar model it's not real.

**Comment by conick** [comment]
Why i cant find this changes and ability to use format option for textinput fields in 2.0.8 version?

**Comment by freezy-sk** [comment]
@conick because that PR #9208 is still not merged

**Comment by lynicidn** [comment]
@cebe why we need format value for text input and area, it will not validate on request, if it validated, than u instantiate not valid model

**Comment by klimov-paul** [comment]
The idea make no sense to me: after number is formatted according to the specific locale (including thousands and decimal separators) it usually becomes unparsable and will no longer recognized as number by JavaScript or PHP.

For example: both for PHP and JavaScript following value is a valid number: `1500200.56`. However after being formatted it becomes following: `1 500 200,56`. Now nethier JavaScript nor PHP is unable to parse this string and extact actual number value from it, unless a specific parser created.

Allowing formatter on text input values will break existing `ActiveForm` client-side validation as well as regular server-side validation.

If you want a text input value been formatted in a specific way like locale-aware decimal number, you should use `maskedinput` jQuery plugin or similar solution.

**Comment by SilverFire** [comment]
PR #9208 is rejected: https://github.com/yiisoft/yii2/pull/9208#issuecomment-199343845

**Comment by lynicidn** [comment]
это бред чистой воды уже, формат это от форматера методы asX , там не только преобразование числовых типов, там есть и asHtml и прочая ненужная фигня.
Если аттрибут модели увас содержит некорректное значение, которое надо преобразовать, то почему вы это делаете в виджете ActiveField, чтобы вы не крутили, придет строка и запишется в валуе строка. Ваша "надуманная проблема" решается куда проще
```php
class MyModel extends Model
{
    public function attributes()
    {
        return ['value'];
    }
    private $_value;
    public function setValue($value)
    {
        $this->_value = $value;
    }

    public function getValue()
    {
        return Yii::$app->formatter->asHtml($this->_value)
    }
}
```

**Comment by samdark** [comment]
Won't be implemented in 2.0.
